library/src/main/java/com/schibsted/spain/barista/interaction/PermissionGranter.kt

```diff
@@ -36,15 +35,43 @@ object PermissionGranter {
         sleepThread(PERMISSIONS_DIALOG_DELAY.toLong())
         val device = UiDevice.getInstance(getInstrumentation())
 
-        val resourceId = if (permissionNeeded == ACCESS_FINE_LOCATION || permissionNeeded == ACCESS_COARSE_LOCATION) {
-            PERMISSIONS_DIALOG_ALLOW_FOREGROUND_ID
-        } else {
-            PERMISSIONS_DIALOG_ALLOW_ID
+        val regex = PERMISSION_DIALOG_ALLOW_FOREGROUND_IDS.joinToString(
+          prefix = "^(",
+          separator = "|",
+          postfix = ")$"
+        ) { it }
+        val allowPermissions = device.findObject(UiSelector()
+            .clickable(true)
+            .checkable(false)
+            .resourceIdMatches(regex)
+        )
+        if (allowPermissions.exists()) {
+          allowPermissions.click()
         }
+      }
+    } catch (e: UiObjectNotFoundException) {
+      Log.e("Barista", "There is no permissions dialog to interact with", e)
+    }
+  }
+
+  @JvmStatic
+  fun allowPermissionOneTime(permissionNeeded: String) {
+    try {
+      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && !hasNeededPermission(getApplicationContext(),
+              permissionNeeded)) {
+        sleepThread(PERMISSIONS_DIALOG_DELAY.toLong())
+        val device = UiDevice.getInstance(getInstrumentation())
+
+        val regex = PERMISSION_DIALOG_ALLOW_ONE_TIME_IDS.joinToString(
+          prefix = "^(",
+          separator = "|",
+          postfix = ")$"
+        ) { it }
```

Comment:
A big part of this code is duplicated in the two methods. It could be extracted.

Reply:
done